core/src/test/java/org/apache/calcite/test/JdbcTest.java

```diff
@@ -2544,35 +2556,117 @@ private void checkNullableTimestamp(CalciteAssert.Config config) {
+              "              final org.apache.calcite.test.JdbcTest.Employee current = (org.apache.calcite.test.JdbcTest.Employee) inputEnumerator.current();\n"
+            + "              final String input_value = current.name;\n"
+            + "              final int input_value0 = current.deptno;\n"
+            + "              Integer case_when_value;\n"
+            + "              if ($L4J$C$org_apache_calcite_runtime_SqlFunctions_eq_) {\n"
+            + "                case_when_value = $L4J$C$Integer_valueOf_1_;\n"
+            + "              } else {\n"
+            + "                case_when_value = (Integer) null;\n"
+            + "              }\n"
+            + "              final Integer binary_call_value1 = case_when_value == null ? (Integer) null : Integer.valueOf(Integer.valueOf(input_value0 * 0) + case_when_value);\n"
+            + "              final String method_call_value = input_value == null || binary_call_value1 == null ? (String) null : org.apache.calcite.runtime.SqlFunctions.substring(input_value, binary_call_value1.intValue());\n"
+            + "              final String trim_value = method_call_value == null ? (String) null : org.apache.calcite.runtime.SqlFunctions.trim(true, true, \" \", method_call_value, true);\n"
+            + "              Integer case_when_value0;\n"
+            + "              if (current.empid > input_value0) {\n"
+            + "                case_when_value0 = $L4J$C$Integer_valueOf_5_;\n"
+            + "              } else {\n"
+            + "                Integer case_when_value1;\n"
+            + "                if (current.deptno * 8 > 8) {\n"
+            + "                  case_when_value1 = $L4J$C$Integer_valueOf_5_;\n"
+            + "                } else {\n"
+            + "                  case_when_value1 = (Integer) null;\n"
+            + "                }\n"
+            + "                case_when_value0 = case_when_value1;\n"
+            + "              }\n"
+            + "              final Integer binary_call_value3 = case_when_value0 == null ? (Integer) null : Integer.valueOf(case_when_value0 - $L4J$C$Integer_valueOf_2_);\n"
+            + "              return trim_value == null || binary_call_value3 == null ? (String) null : org.apache.calcite.runtime.SqlFunctions.substring(trim_value, binary_call_value3.intValue());"
+        )
         .returns("T=ll\n"
             + "T=ic\n"
             + "T=bastian\n"
             + "T=eodore\n");
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3142">[CALCITE-3142]
+   * An NPE when rounding a nullable numeric</a>. */
+  @Test public void testRoundingNPE() {
+    CalciteAssert.that()
+        .query("SELECT ROUND(CAST((X/Y) AS NUMERIC), 2) "
+            + "FROM (VALUES (1, 2), (NULLIF(5, 5), NULLIF(5, 5))) A(X, Y)")
+        .planContains("      final Object[] current = (Object[]) inputEnumerator.current();\n"
```

Comment:
Is there a particular reason to verify that the generated code is exactly the one included in the test? I would say that for this test it suffices to verify that the query runs correctly without NPE and not really checking the code line by line.

Reply:
No particular reason, just provide the generated code as @vlsi 's request.

Reply:
There are at least two cases for verifying the generated code:
1) Ensure the optimizations are there
2) Ensure the code is safe (e.g. to check it does not dereference before checking for null)

We don't have "performance regression" tests yet, so it makes sense to add assertions for the generated source.

Reply:
Code verification removed.

Reply:
@vlsi I totally agree on you on the fact that we should verify the generated code in some cases, especially when it has to do with optimizations. 

On the other hand, I would not put assertions with the generated code in every query. I've seen that @DonnyZone removed a few code assertions(thanks) and I prefer it this way. If you think that those were important we can add them back later before the final merge.

Reply:
Essential assertions in `RexCall`'s implementors are preserved. We just remove some `Assert`/`AssertionError`/`AlwaysNull` in the framework, because they are introduced by the current algorithm which makes `RexNode`'s nullable state immutable. 

Technically speaking, the nullable state of `RexNode` can be immutable during codegen. In the new one, each `RexNode` is visited only once. From its type, we can know whether it is nullable directly, without maintaining any states. Therefore, many checks (e.g., `isNullable(...)`) are not necessary again.

To conclude, the following things are extra and make the progress complicated. We removed them directly in the new one.

1. `private final Map<? extends RexNode, Boolean> exprNullableMap` in `RexToLixTranslator`. Each `RexNode` is visited only once, we do not need it again.
2. `private final RexToLixTranslator parent` in `RexToLixTranslator`. We do not need to recursively get a `RexNode`'s nullable state again.
3. `AlwaysNull` exception. `AlwaysNull` in current codegen is the root cause of some underyling problems. It buries some real problems and brings 'magic/undeterministic' result. From my personal point of view, I'm not in favour of `AlwaysNull`. According to document, it is proposed for 'unusual' situation. Since a `RexNode` is legal, the code generation progress should not step into such situation. What's worse is, the exception is catched and returns `null` aggressively in some places.

> Thrown in the unusual (but not erroneous) situation where the expression
> we are translating is the null literal but we have already checked that
> it is not null.